### PR TITLE
Add AggLet node (required for MatrixAggregateColsByKey lowering)

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 1]
```diff
@@ -134,8 +134,10 @@ object Simplify {
     case IsNA(x) if isDefinitelyDefined(x) => False()
 
     case Let(n1, v, Ref(n2, _)) if n1 == n2 => v
+    case AggLet(n1, v, Ref(n2, _)) if n1 == n2 => v
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: I think this is wrong.  n2 is referring to something in env, but AggLet is binding something in aggEnv.

File: hail/src/main/scala/is/hail/expr/ir/Subst.scala [hunk 2]
```diff
@@ -14,7 +14,10 @@ object Subst {
         env.lookupOption(name).getOrElse(x)
       case Let(name, v, body) =>
         val newv = subst(v)
-        Let(name, newv, subst(body, env.delete(name).bind(name, Ref(name, newv.typ))))
+        Let(name, newv, subst(body, env.bind(name, Ref(name, newv.typ))))
+      case AggLet(name, v, body) =>
+        val newv = subst(v)
+        Let(name, newv, subst(body, aggEnv = aggEnv.bind(name, Ref(name, newv.typ))))
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/Subst.scala]: This should generate an AggLet.

**Comment by tpoterba** [comment]
I'm also missing agglet extraction logic in StagedExtractAggs

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: it should always be false?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: are the agg binding rules below wrong?

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: No, there could be shadowing: `(Let a v (AggLet a w (Ref a))` is correct and the ref should refer to the (first) let.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: ok, removed these rules.

**Pull request merged**